source/extensions/transport_sockets/tcp_stats/tcp_stats.cc

```diff
@@ -145,14 +152,72 @@ void TcpStatsSocket::recordStats() {
                  tcp_info->tcpi_bytes_received);
   update_counter(config_->stats_.cx_tx_bytes_sent_, last_cx_tx_bytes_sent_,
                  tcp_info->tcpi_bytes_sent);
+  update_counter(config_->stats_.cx_data_segments_delivered_, last_cx_data_segments_delivered_,
+                 tcp_info->tcpi_delivered);
+  update_counter(config_->stats_.cx_reordering_, last_cx_reordering_, tcp_info->tcpi_reordering);
 
   update_gauge(config_->stats_.cx_tx_unsent_bytes_, last_cx_tx_unsent_bytes_,
                tcp_info->tcpi_notsent_bytes);
   update_gauge(config_->stats_.cx_tx_unacked_segments_, last_cx_tx_unacked_segments_,
                tcp_info->tcpi_unacked);
+  update_gauge(config_->stats_.cx_rto_us_, last_cx_rto_us_, tcp_info->tcpi_rto);
+  update_gauge(config_->stats_.cx_ato_us_, last_cx_ato_us_, tcp_info->tcpi_ato);
+  update_gauge(config_->stats_.cx_lost_, last_cx_lost_, tcp_info->tcpi_lost);
+  update_gauge(config_->stats_.cx_tx_ssthreshold_, last_cx_tx_ssthreshold_,
+               tcp_info->tcpi_snd_ssthresh);
+  update_gauge(config_->stats_.cx_rx_ssthreshold_, last_cx_rx_ssthreshold_,
+               tcp_info->tcpi_rcv_ssthresh);
+  update_gauge(config_->stats_.cx_tx_mss_bytes_, last_cx_tx_mss_bytes_, tcp_info->tcpi_snd_mss);
+  update_gauge(config_->stats_.cx_rx_mss_bytes_, last_cx_rx_mss_bytes_, tcp_info->tcpi_rcv_mss);
+  update_gauge(config_->stats_.cx_advmss_bytes_, last_cx_advmss_bytes_, tcp_info->tcpi_advmss);
+  update_gauge(config_->stats_.cx_pmtu_bytes_, last_cx_pmtu_bytes_, tcp_info->tcpi_pmtu);
 
   config_->stats_.cx_rtt_us_.recordValue(tcp_info->tcpi_rtt);
   config_->stats_.cx_rtt_variance_us_.recordValue(tcp_info->tcpi_rttvar);
+  config_->stats_.cx_congestion_window_.recordValue(snd_cwnd);
+  config_->stats_.cx_pacing_rate_.recordValue(pacing_rate_mb);
+  config_->stats_.cx_delivery_rate_.recordValue(delivery_rate_mb);
+  config_->stats_.cx_rcv_rtt_.recordValue(tcp_info->tcpi_rcv_rtt);
+  config_->stats_.cx_tx_window_scale_.recordValue(tcp_info->tcpi_snd_wscale);
+  config_->stats_.cx_rx_window_scale_.recordValue(tcp_info->tcpi_rcv_wscale);
+
+  Envoy::ProtobufWkt::Struct meta_struct;
+  auto& fields = *meta_struct.mutable_fields();
+  *fields["cx_tx_segments"].mutable_string_value() = absl::StrCat(tcp_info->tcpi_segs_out);
```

Comment:
Not sure this is a case we want to support, probably need feedback from the maintainer.

I'm also thinking if this is really a case we want to support, then whether we have a generic way to get stats for the access log, just like %METADATA(...)%, we can add %STATS(...)%

Reply:
I don't think we should populate all this metadata by default. Maybe consider it behind a config flag, but I'd rather have that be a separate issue/PR and we can discuss it there.

If we do populate the metadata, does it make sense to do it periodically, or only at the end of the connection?

Reply:
Thanks for quick review and sorry for the delay in addressing the comments.
I have created following issue to get better way to access these tcp_info stats in access logger.
https://github.com/envoyproxy/envoy/issues/27391